<a href="https://colab.research.google.com/github/ivanachen1/LinearRegression/blob/main/Linear_Reggresion_AutoIndustries.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime

In [2]:
%cd /content/drive/MyDrive/Colab Notebooks/Machine Learning/

/content/drive/MyDrive/Colab Notebooks/Machine Learning


In [3]:
import LinearRegressionAux as aux

In [4]:
dataframe = aux.carga_datos()
dataframe = dataframe.dropna(axis= 0)


In [5]:
dataframe = aux.cambio_formatos(dataframe)
total_articulos = dataframe["product_id"].unique()
len(total_articulos)

5774

In [6]:
lista_anios = dataframe["anio"].unique()

In [7]:
#Analizo las ventas de cada articulo los ultimos 3 años
pivot_table = dataframe.pivot_table(dataframe,index= "product_id",columns="anio",aggfunc= {"cantidad":"sum"},fill_value=0)

In [8]:
pivot_table.columns = pivot_table.columns.droplevel(0)
pivot_table.columns.name = None              
pivot_table = pivot_table.reset_index().rename_axis(None,axis = 1) 
pivot_table

product_id  2018  2019  2020
0           3743     5    34    24
1           3744     1     3     2
2           3747     4    18    14
3           3748  1010  4777  3070
4           3749   264  1045   642
...          ...   ...   ...   ...
5769       13016     0     7     1
5770       13018     0     2     0
5771       13020     0     1     1
5772       13021     0     3     0
5773       13022     0     1     0

[5774 rows x 4 columns]

In [9]:
today = datetime.today()
today = (today.year) - 1

In [10]:
data_filtrada = pivot_table[pivot_table[today] == 0]
data_filtrada = pivot_table[pivot_table[today-1] == 0]
data_filtrada
articulos_sin_ventas_ultimos_dos_anios = data_filtrada["product_id"].unique()
len(articulos_sin_ventas_ultimos_dos_anios)

160

In [11]:
articulos_sin_ventas = data_filtrada["product_id"].unique()
articulos_sin_ventas

array([ 3778,  3833,  3873,  5141,  5171,  5301,  5302,  5355,  5415,
        5432,  5450,  5479,  5491,  5495,  5578,  5813,  5874,  6151,
        6168,  6170,  6294,  6330,  6350,  6610,  6634,  6637,  6645,
        6657,  6663,  6723,  6835,  6853,  6883,  6933,  7006,  7009,
        7043,  7130,  7346,  7381,  7473,  7494,  7566,  7590,  7686,
        7687,  7747,  7828,  7837,  7960,  7973,  7984,  8115,  8116,
        8125,  8214,  8290,  8436,  8518,  8528,  8545,  8642,  8649,
        8655,  8656,  8658,  8696,  8749,  8769,  8783,  8803,  8812,
        8815,  8876,  8940,  9016,  9057,  9120,  9132,  9134,  9143,
        9181,  9195,  9197,  9218,  9409,  9421,  9482,  9559,  9650,
        9730,  9751,  9976, 10080, 10237, 10412, 10415, 10447, 10489,
       10531, 10544, 10678, 10753, 10787, 10825, 10852, 10861, 10908,
       10945, 10967, 10978, 10979, 11070, 11136, 11205, 11208, 11276,
       11382, 11395, 11434, 11477, 11702, 11742, 11773, 11777, 11791,
       11868, 11871,

Quito del Dataframe los articulos que no se vendieron en los ultimos 2 años

In [12]:
#filtro el dataset
dataframe = dataframe[~ dataframe["product_id"].isin(articulos_sin_ventas)]
dataframe

operating_unit_id  product_id     fecha  mes  anio  cantidad   tipo
0                      4        3748  01/09/19    9  2019         3  Venta
1                      4        3752  01/09/19    9  2019         1  Venta
2                      4        3775  01/09/19    9  2019         1  Venta
3                      4        3777  01/09/19    9  2019         1  Venta
4                      4        3781  01/09/19    9  2019         1  Venta
...                  ...         ...       ...  ...   ...       ...    ...
73640                 94       12599  01/02/20    2  2020         1  Venta
73641                 94       12619  01/02/20    2  2020         1  Venta
73642                 94       12730  01/02/20    2  2020         4  Venta
73643                 94       12763  01/02/20    2  2020         1  Venta
73644                 94       12957  01/02/20    2  2020         4  Venta

[841067 rows x 7 columns]

In [13]:
pivot_table = dataframe.pivot_table(dataframe,index= "product_id",columns = "anio",aggfunc= {"cantidad":"sum"},fill_value=0)
pivot_table.columns = pivot_table.columns.droplevel(0)
pivot_table.columns.name = None              
pivot_table = pivot_table.reset_index().rename_axis(None,axis = 1) 
pivot_table

product_id  2018  2019  2020
0           3743     5    34    24
1           3744     1     3     2
2           3747     4    18    14
3           3748  1010  4777  3070
4           3749   264  1045   642
...          ...   ...   ...   ...
5609       13016     0     7     1
5610       13018     0     2     0
5611       13020     0     1     1
5612       13021     0     3     0
5613       13022     0     1     0

[5614 rows x 4 columns]

In [14]:
n = len(pivot_table["product_id"].unique())
lista_articulos_vendidos = []

for i in [10,20,30,40,50,60,70,80,90,100,110,200,250,500,1000,2000,3000,5000,6000,9000,4000000]:
  data = pivot_table[pivot_table[2020] < i]
  articulos = len(data["product_id"].unique())
  list_articulos = data["product_id"].unique()
  lista_articulos_vendidos.append((i,list_articulos))
  print("Los articulos que poseen menos de {} ventas en 2020 son {}, representando el {}% de articulos con ventas en los ultimos años".format(i,articulos,round(articulos / n * 100,2)))

Los articulos que poseen menos de 10 ventas en 2020 son 2568, representando el 45.74% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 20 ventas en 2020 son 3319, representando el 59.12% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 30 ventas en 2020 son 3720, representando el 66.26% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 40 ventas en 2020 son 3992, representando el 71.11% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 50 ventas en 2020 son 4209, representando el 74.97% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 60 ventas en 2020 son 4372, representando el 77.88% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 70 ventas en 2020 son 4515, representando el 80.42% de articulos con ventas en los ultimos años
Los articulos que poseen menos de 80 ventas en 2020 son 4631, representando el 82.49% de a

In [15]:
pivot_table = dataframe.pivot_table(dataframe,index= "product_id",columns = "anio",aggfunc= {"cantidad":"sum"},fill_value=0)
pivot_table.columns = pivot_table.columns.droplevel(0)
pivot_table.columns.name = None              
pivot_table = pivot_table.reset_index().rename_axis(None,axis = 1) 
pivot_table

product_id  2018  2019  2020
0           3743     5    34    24
1           3744     1     3     2
2           3747     4    18    14
3           3748  1010  4777  3070
4           3749   264  1045   642
...          ...   ...   ...   ...
5609       13016     0     7     1
5610       13018     0     2     0
5611       13020     0     1     1
5612       13021     0     3     0
5613       13022     0     1     0

[5614 rows x 4 columns]

In [16]:
dataframe,lista_anios = aux.campos_datetime(dataframe)
dataframe

/content/drive/MyDrive/Colab Notebooks/Machine Learning/LinearRegressionAux.py:53: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe.sort_values(by = "Fecha_format_datetime",inplace = True)
/content/drive/MyDrive/Colab Notebooks/Machine Learning/LinearRegressionAux.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe["day"] = pd.DatetimeIndex(dataframe["Fecha_format_datetime"]).day
/content/drive/MyDrive/Colab Notebooks/Machine Learning/LinearRegressionAux.py:56: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a Data

operating_unit_id  product_id  cantidad  ...    week2  Quarter  Semester
48078                  93       13000         4  ...  2018_01        1         1
40087                  94        9401         1  ...  2018_01        1         1
47052                  91        9835         4  ...  2018_01        1         1
47053                  91        9864         1  ...  2018_01        1         1
47054                  91        9895         2  ...  2018_01        1         1
...                   ...         ...       ...  ...      ...      ...       ...
72871                  90       10134         1  ...  2020_49        4         2
105438                 89        8142         1  ...  2020_49        4         2
72864                  90        8920         2  ...  2020_49        4         2
72824                  90        5467         2  ...  2020_49        4         2
72841                  90        7214         3  ...  2020_49        4         2

[840764 rows x 11 columns]

In [17]:
lista_anios

array([2018, 2019, 2020])

In [18]:
dataframe = aux.crear_accumulative_week(dataframe,lista_anios)
dataframe

operating_unit_id  product_id  ...  Semester acumulative_week
48078                  93       13000  ...         1                2
40087                  94        9401  ...         1                2
47052                  91        9835  ...         1                2
47053                  91        9864  ...         1                2
47054                  91        9895  ...         1                2
...                   ...         ...  ...       ...              ...
72871                  90       10134  ...         2              204
105438                 89        8142  ...         2              204
72864                  90        8920  ...         2              204
72824                  90        5467  ...         2              204
72841                  90        7214  ...         2              204

[840764 rows x 12 columns]

In [19]:
# filtro semestral dataframe
#voy a obtener la lista de articulos que quiero excluir por que no se vendieron en el ultimo semestre
data = dataframe
data = data.pivot_table(data,index=["product_id","Semester"],columns=["year"],aggfunc= {"cantidad":"sum"},fill_value=0)
data.columns = data.columns.droplevel(0) 
data.columns.name = None             
data = data.reset_index().rename_axis(None,axis = 1)
data

product_id  Semester  2018  2019  2020
0            3743         1     2    23    21
1            3743         2     3    11     3
2            3744         1     0     2     2
3            3744         2     1     1     0
4            3747         1     0     9     9
...           ...       ...   ...   ...   ...
10256       13016         2     0     7     1
10257       13018         2     0     2     0
10258       13020         2     0     1     1
10259       13021         2     0     3     0
10260       13022         2     0     1     0

[10261 rows x 5 columns]

In [20]:
data_con_ventas_ultimo_semestre = data[data["Semester"] == 2]
data_con_ventas_ultimo_semestre = data_con_ventas_ultimo_semestre[data_con_ventas_ultimo_semestre[2020] > 20]
data_con_ventas_ultimo_semestre

product_id  Semester  2018  2019  2020
7            3748         2   881  2480   974
9            3749         2   216   556   223
11           3750         2    81   184    59
13           3751         2    57    93    29
15           3752         2    63   112    40
...           ...       ...   ...   ...   ...
10215       12984         2     0    24    24
10219       12986         2    46    94    39
10234       12996         2    39   113    52
10236       12997         2    10    30    21
10238       12998         2     0    57    55

[1254 rows x 5 columns]

In [21]:
productos_vendidos = data_con_ventas_ultimo_semestre["product_id"].unique()
len(productos_vendidos)

1254

Conclusion

En base a lo analizado, voy a crear modelos con los articulos que tuvieron mas de 20 ventas, desestimando a los anteriores

In [22]:
#dataframe = dataframe.pivot_table(dataframe,index=["week2","product_id"],aggfunc= {"cantidad":"sum"},fill_value=0)
#dataframe.columns.name = None              
#dataframe = dataframe.reset_index().rename_axis(None,axis = 1) 


# tengo que meter las 52 semanas de todos los años

lista_numeros_semana = list(dataframe["acumulative_week"].unique())
lista_weeks_ordenadas = []

for i in range(1,len(lista_numeros_semana)+1,1):
  if i in lista_numeros_semana:
    continue
  else:
    lista_numeros_semana.append(i)


In [23]:
dataframe_to_predict = aux.agregar_semanas_cantidades_cero(dataframe,productos_vendidos,print_resultados= False)

lista_numeros_semana
[2, 6, 7, 10, 11, 15, 19, 23, 24, 28, 32, 35, 36, 37, 38, 39, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 1, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 194, 192, 196, 202, 204, 3, 4, 5, 8, 9, 12, 13, 14, 16, 17, 18, 20, 21, 22, 25, 26, 27, 29, 30, 31, 33, 34, 40, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 188, 189, 190, 191, 193, 195, 197, 198, 199, 200, 201, 203]
lista original
[2, 6, 7, 10, 11, 15, 19, 23, 24, 28, 32, 35, 36, 3

In [24]:
dataframe_to_predict

product_id  cantidad  acumulative_week
46894        3748         1                 2
47878        3748         1                 2
41457        3748         4                 2
47304        3748         1                 2
47622        3748         1                 2
...           ...       ...               ...
174         12998         0               198
175         12998         0               199
176         12998         0               200
177         12998         0               201
178         12998         0               203

[683462 rows x 3 columns]

In [31]:
#uno = dataframe_to_predict.pivot_table(index="product_id",columns= "acumulative_week",aggfunc= {"cantidad":"sum"},fill_value=0)
#uno[uno["product_id"] == 3750]

pivot_table = dataframe_to_predict.pivot_table(index="product_id",columns= "acumulative_week",aggfunc= {"cantidad":"sum"},fill_value=0)
pivot_table.columns = pivot_table.columns.droplevel(0)
pivot_table.columns.name = None              
pivot_table = pivot_table.reset_index().rename_axis(None,axis = 1)
pivot_table 
#pivot_table[pivot_table["product_id"] == 3748]

product_id   1   2  3  4  5   6  ...  198  199  200  201  202  203  204
0           3748  14  20  0  0  0  16  ...    0    0    0    0   80    0   23
1           3749   6  10  0  0  0   3  ...    0    0    0    0   21    0    1
2           3750   1   3  0  0  0   3  ...    0    0    0    0    3    0    2
3           3751   0   3  0  0  0   1  ...    0    0    0    0    0    0    1
4           3752   1   3  0  0  0   0  ...    0    0    0    0    0    0    1
...          ...  ..  .. .. .. ..  ..  ...  ...  ...  ...  ...  ...  ...  ...
1249       12984   0   0  0  0  0   0  ...    0    0    0    0    0    0    0
1250       12986   0   1  0  0  0   0  ...    0    0    0    0    4    0    1
1251       12996   0   0  0  0  0   0  ...    0    0    0    0    0    0    2
1252       12997   0   1  0  0  0   0  ...    0    0    0    0    0    0    1
1253       12998   0   0  0  0  0   0  ...    0    0    0    0    0    0    4

[1254 rows x 205 columns]

ANOTACION:

Hasta este punto tengo toda la data limpia en el dataframe_to_predict. Se saco la conclusion de que muy probablemente no sirva crear una regresion linear articulo, tienda sino que es mejor crear por articulo

In [26]:
import statsmodels.api as sm
import numpy as np
# voy a trabajar con un solo articulo y luego lo replico a los demas

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [27]:
data = aux.predecir(dataframe_to_predict,productos_vendidos)
data

Articulo  R square
0         3748  0.463108
1         3749  0.671295
2         3750  0.731764
3         3751  0.751061
4         3752  0.687691
...        ...       ...
1249     12984  0.592992
1250     12986  0.723029
1251     12996  0.786634
1252     12997  0.789601
1253     12998  0.151912

[1254 rows x 2 columns]

In [32]:
R2_menor_50 = 0
R2_beetween_5070 = 0
R2_more_70 = 0
for row in data.itertuples():
  if row[2] >= 0 and row[2] < 0.50:
    R2_menor_50 += 1
  elif row[2] >= 0.50 and row[2] < 0.70:
    R2_beetween_5070 += 1
  else:
    R2_more_70 += 1

print("La cantidad de articulos con R2 menor a 0.50 es ",R2_menor_50)
print("La cantidad de articulos con R2 mayor a 0.50 y menor a 0.70 es ",R2_beetween_5070)
print("La cantidad de articulos con R2 mayor a 0.70 es",R2_more_70)

La cantidad de articulos con R2 menor a 0.50 es  0
La cantidad de articulos con R2 mayor a 0.50 y menor a 0.70 es  0
La cantidad de articulos con R2 mayor a 0.70 es 814


In [29]:
data = data[data["R square"] > 0.70]
data

Articulo  R square
2         3750  0.731764
3         3751  0.751061
11        3784  0.741800
13        3792  0.787265
15        3802  0.727900
...        ...       ...
1247     12982  0.766040
1248     12983  0.749634
1250     12986  0.723029
1251     12996  0.786634
1252     12997  0.789601

[814 rows x 2 columns]

https://www.statsmodels.org/dev/example_formulas.html

Es que me arme en base al modelo, una prediccion y podria crear un array con las mismas y enchufarlo en el dataframe to predict para ver las diferencias

Puedo crear un funcion que me calcule todos los modelos y me entregue una tabla con:



1.   Todos los R2
2.   Los articulos de esos R2

Podria crear un filtro que si el R2 es mayor a 0.70 me haga las predicciones y luego las comparo con el dataframe historico.













Conclusiones del analisis

1) 160 Articulos no tuvieron ventas los ultimos 2 años

2) 2728 articulos tuvieron menos de 10 ventas en 2020

2) 3891 articulos no se vendieron en los ultimos 6 meses y 1254 si fueron vendidos

3) 1228 articulos tuvieron ventas en los ultimos años  y en los ultimos 6 meses

4) De estos 1228 articulos, 1056 vendieron 10 articulos en el año representando el 85,75%. 

5) En el histograma esta la participacion de la cantidad de articulos vendidos los ultimos años y meses sobre el total

nota = tomar las semanas, 

Conclusiones

Despues de limpiar mucho los datos, y agruparlos por semana, tengo que comprobar si vale la pena hacer regresiones lineales de los datos ya que tengo muchos articulos que en muchas semanas no tienen ventas. Lo que se podria evaluar es el nivel de ventas por articulo por mes para analizar cuales son posibles candidatos a predecir

Voy a excluir los articulos que no se vendieron en los ultimos 6 meses